### PR TITLE
[Java8] Replace ConstantPoolTypeIntrospector

### PR DESCRIPTION
**Pull request by mpkorstanje**
## Summary

Replace ConstantPoolTypeIntrospector with [TypeResolver](https://github.com/jhalterman/typetools) and refactor all the Java8 related stuff out of cucumber-java.

## Motivation and Context

To determine which argument types a lambda function requires we created a [ConstantPoolTypeIntrospector](https://github.com/cucumber/cucumber-jvm/blob/aa4c9a012d623cbbedeaa0e54af7c30f32825928/java8/src/main/java/cucumber/runtime/java8/ConstantPoolTypeIntrospector.java). However it has never functioned quite correctly (#937, #1140, [#957](https://github.com/cucumber/cucumber-jvm/issues/957#issuecomment-315467655)), was prone to breaking (#912, #914)  and hasn't been tested much (#1048).

It is important to understand that while we will get a properly functioning and tested replacement, TypeResolver uses the same ConstantPool and thus has the same potential to break. However because TypeResolver is used by a much larger audience I expect these problems to be shallow.

Because this changes the interface of Java8StepDefinition it made sense to refactor all the Java8 related stuff out of cucumber-java. This will make it easier in the future to add things like KotlinStepDefintions without creating a separate KotlinBackend.

## How Has This Been Tested?

Check if the automated tests were still passing. 

## Types of changes
 
- [X] Bug fix (non-breaking change which fixes an issue).
     -  Method references that use primitives are correctly understood.
- [ ] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).
     -  When using anonymous innerclasses with Java8 Stepdefinitions it was possible to use typed lists and maps as an argument. This is no longer possible as TypeResolver only returns classes, not the full type.
    - In fact it will not be possible to use anonymous inner classes as step definitions without java8 at all.

## Checklist:

- [X] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: java/pom.xml [hunk 1]
```diff
@@ -17,16 +17,6 @@
...
         </dependency>
-        <dependency>
-            <groupId>info.cukes</groupId>
-            <artifactId>cucumber-jvm-deps</artifactId>
-            <scope>provided</scope>
-        </dependency>
-        <dependency>
-            <groupId>io.cucumber</groupId>
-            <artifactId>gherkin</artifactId>
-            <scope>provided</scope>
```
Comment (mpkorstanje) [inline java/pom.xml]: Actually both are transitive dependency through cucumber-core.

File: java/src/main/java/cucumber/runtime/java/JavaBackend.java [hunk 2]
```diff
@@ -59,24 +60,25 @@ private Snippet createSnippet() {
...
+                    currentThread().getContextClassLoader()),
+                currentThread().getContextClassLoader()));
     }
 
     public JavaBackend(ObjectFactory objectFactory, ClassFinder classFinder) {
-        this.objectFactory = objectFactory;
         this.classFinder = classFinder;
-        methodScanner = new MethodScanner(classFinder);
+        this.objectFactory = objectFactory;
+        this.methodScanner = new MethodScanner(classFinder);
```
Comment (mpkorstanje) [inline java/src/main/java/cucumber/runtime/java/JavaBackend.java]: Rewrote the constructors to be idiomatic java.

File: java/src/main/java/cucumber/runtime/java/JavaBackend.java [hunk 3]
```diff
@@ -157,44 +159,30 @@ void addStepDefinition(Annotation annotation, Method method) {
         }
     }
 
-    public void addStepDefinition(String regexp, long timeoutMillis, StepdefBody body, TypeIntrospector typeIntrospector) {
-        try {
-            glue.addStepDefinition(new Java8StepDefinition(Pattern.compile(regexp), timeoutMillis, body, typeIntrospector));
```
Comment (mpkorstanje) [inline java/src/main/java/cucumber/runtime/java/JavaBackend.java]: This has moved into the generated code.

File: java/src/main/java/cucumber/runtime/java/JavaBackend.java [hunk 4]
```diff
@@ -157,44 +159,30 @@ void addStepDefinition(Annotation annotation, Method method) {
...
                 String[] tagExpressions = ((After) annotation).value();
                 long timeout = ((After) annotation).timeout();
-                glue.addAfterHook(new JavaHookDefinition(method, tagExpressions, ((After) annotation).order(), timeout, objectFactory));
+                addAfterHookDefinition(new JavaHookDefinition(method, tagExpressions, ((After) annotation).order(), timeout, objectFactory));
             }
         }
     }
 
-    public void addBeforeHookDefinition(String[] tagExpressions, long timeoutMillis, int order, HookBody body) {
-        glue.addBeforeHook(new Java8HookDefinition(tagExpressions, order, timeoutMillis, body));
```
Comment (mpkorstanje) [inline java/src/main/java/cucumber/runtime/java/JavaBackend.java]: This has moved into the generated code.

File: java/src/main/java/cucumber/runtime/java/MethodScanner.java [hunk 5]
```diff
@@ -18,7 +18,7 @@
 
     private final ClassFinder classFinder;
 
-    public MethodScanner(ClassFinder classFinder) {
+    MethodScanner(ClassFinder classFinder) {
```
Comment (mpkorstanje) [inline java/src/main/java/cucumber/runtime/java/MethodScanner.java]: Unrelated cleanup.

File: java/src/test/java/cucumber/runtime/java/java8test/AnonInnerClassStepdefs.java [hunk 6]
```diff
@@ -1,20 +0,0 @@
-package cucumber.runtime.java.java8test;
-
-import cucumber.api.java8.GlueBase;
-import cucumber.api.java8.StepdefBody;
-import cucumber.runtime.java.JavaBackend;
-
-import static org.junit.Assert.assertEquals;
-
-public class AnonInnerClassStepdefs implements GlueBase {
```
Comment (mpkorstanje) [inline java/src/test/java/cucumber/runtime/java/java8test/AnonInnerClassStepdefs.java]: moved to cucumber.runtime.java8.test.AnonInnerClassStepdefs

File: java8/src/main/code_generator/I18n.java8.txt [hunk 7]
```diff
@@ -1,28 +1,37 @@
...
-        JavaBackend.INSTANCE.get().addStepDefinition(regexp, 0, body, ConstantPoolTypeIntrospector.INSTANCE);
+        JavaBackend.INSTANCE.get().addStepDefinition(
+            new Java8StepDefinition(regexp, 0, StepdefBody.A0.class, body)
+        );
     }
 
     default void ${java.text.Normalizer.normalize(kw.replaceAll("[\\s',!]", ""), java.text.Normalizer.Form.NFC)}(final String regexp, final long timeoutMillis, final StepdefBody.A0 body) {
-        JavaBackend.INSTANCE.get().addStepDefinition(regexp, timeoutMillis, body, ConstantPoolTypeIntrospector.INSTANCE);
+        JavaBackend.INSTANCE.get().addStepDefinition(
+            new Java8StepDefinition(regexp, timeoutMillis, StepdefBody.A0.class, body)
```
Comment (mpkorstanje) [inline java8/src/main/code_generator/I18n.java8.txt]: Typetools resolves the arguments of the the generic type (StepdefBody.A1) using the type variable information from body (the implementing class) so we pass these in here.

File: java8/src/test/java/cucumber/runtime/java8/test/LambdaStepdefs.java [hunk 8]
```diff
@@ -65,19 +66,24 @@ public LambdaStepdefs() {
...
     }
 
     private void methodThatDeclaresException() throws Throwable {
     }
     private void methodWithAnArgument(Integer cuckes) throws Throwable {
         assertEquals(42, cuckes.intValue());
     }
+    private void methodWithAnIntArgument(int cuckes) throws Throwable {
+        assertEquals(42, cuckes);
+    }
```
Comment (mpkorstanje) [inline java8/src/test/java/cucumber/runtime/java8/test/LambdaStepdefs.java]: Fixed by using TypeResolver

File: java/src/main/java/cucumber/runtime/java/JavaBackend.java [hunk 9]
```diff
@@ -59,24 +60,25 @@ private Snippet createSnippet() {
...
 
     public JavaBackend(ObjectFactory objectFactory) {
-        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-        ResourceLoader resourceLoader = new MultiLoader(classLoader);
-        classFinder = new ResourceLoaderClassFinder(resourceLoader, classLoader);
-        methodScanner = new MethodScanner(classFinder);
-        this.objectFactory = objectFactory;
+        this(objectFactory,
+            new ResourceLoaderClassFinder(
+                new MultiLoader(
```
Comment (brasmusson) [inline java/src/main/java/cucumber/runtime/java/JavaBackend.java]: I'm not sure all these line breaks improve readability, especially 
`new MultiLoader(currentThread().getContextClassLoader())` seems to fit better on one line.

File: java8/src/main/java/cucumber/runtime/java8/Java8HookDefinition.java [hunk 10]
```diff
@@ -1,17 +1,17 @@
-package cucumber.runtime.java;
+package cucumber.runtime.java8;
+
+import static java.util.Arrays.asList;
```
Comment (brasmusson) [inline java8/src/main/java/cucumber/runtime/java8/Java8HookDefinition.java]: I've found that the most common ordering of import in the code base is static imports after the other imports, so I've added that ordering to the Eclipse configuration I use for Cucumber-JVM, but you seems to have another configuration.

**Comment by mpkorstanje** [comment]
I am interested in your opinion about separating out the java8 bits from cucumber-java. Moving it out of the cucumber-java, out of the JavaBackend, means that any information the Java8StepDefintion uses will have to be either locally available in the generated steps globally accessible.

With the upcoming CucumberExpressions this might be a problem. CucumberExpressions need the the TypeRegistry to be instantiated.

File: java8/src/main/java/cucumber/runtime/java8/Java8HookDefinition.java (hunk 10 above)
Comment (mpkorstanje) [inline java8/src/main/java/cucumber/runtime/java8/Java8HookDefinition.java]: I've been using the default that comes with IDEA. What does Eclipse do by default? I'd prefer a minimal amount of configuration if possible.

Comment (mpkorstanje) [inline java8/src/main/java/cucumber/runtime/java8/Java8HookDefinition.java]: Looks like I have the ability to quickly format the imports of the entire project. We can pick a standard now and adhere to it.

https://www.jetbrains.com/help/idea/optimizing-imports.html

File: java/src/main/java/cucumber/runtime/java/JavaBackend.java (hunk 9 above)
Comment (mpkorstanje) [inline java/src/main/java/cucumber/runtime/java/JavaBackend.java]: You are right. I was missing a constructor `JavaBackend(ObjectFactory, ResourceLoader)`. This made look into the root cause, why do we have these constructors in the first place. It turns out they're only used by tests which needed them because they were poorly written. 

I've added a commit which improves these test somewhat.

**Comment by mpkorstanje** [comment]
I have a solution for the CucumberExpressions in mind. Rather then adding the Java8StepDefintion directly to the backend:

```java
JavaBackend.INSTANCE.get().addStepDefinition(
     new Java8StepDefinition(regexp, timeoutMillis, StepdefBody.A0.class, body)
);
```
We'd provide a Function<TypeRegistry,StepDefinition> (or the java7 equivalent) that will instantiate one:

```java
JavaBackend.INSTANCE.get().addStepDefinition((typeRegistry) -> 
     new Java8StepDefinition(typeRegistry, regexp, timeoutMillis, StepdefBody.A0.class, body)
);
```
